### PR TITLE
[py] Batch Execution Interface

### PR DESCRIPTION
**Pull request by Ankush-lastmile**
[py] Batch Execution Interface

## What

Starting point to create a batch execution interface for AIConfig. This diff creates the interface in python. We can implement the typescript equivalent on top.

Methods Implemented:
- `config.py` (AIConfigRuntime)
- - run_batch() calls `run_batch` of the model parser
- `model_parser.py` (abstract model parser class)
- - run_batch()

## Usage
User will need to construct a list of params for batch, pass it into the `run_batch` method. Very similar to the `run()` method.

### Output type
The resulting object of a Batch Execute will contain the Execute Result, and the resolved completion params with paramters. It would look smoething like this:

```
[
  (
    ExecuteResult(), aiconfig.resolve()
   ),
  (
    ExecuteResult(), aiconfig.resolve()
   )
]
```

Check out the implementation for more details, an example is provided in the docstring

Discussed interface design with @saqadri. We can iterate on top of this but this should get us started.

### Note:

This diff does not implement concurrent request batching. One things would be required on top of this diff for a robust batch exec to be implemented in the model parsers.
1. Exponential Backoff inside model parser implementations at request time. For ex: [OpenAI Backoff impl](https://platform.openai.com/docs/guides/rate-limits/error-mitigation?context=tier-free#:~:text=Example%203%3A-,Manual%20backoff%20implementation,-If%20you%20don%27t)

## Test Plan

Created a config which has a parameterized dependency chain. Batch executed on a list of params

Draft PR containing test plan: [#472](https://github.com/lastmile-ai/aiconfig/pull/472)

https://github.com/lastmile-ai/aiconfig/assets/141073967/e071361f-637f-442b-86c5-edec68b5eba7

This is for issue [#401](https://github.com/lastmile-ai/aiconfig/issues/401)

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/lastmile-ai/aiconfig/pull/469).
* #472
* __->__ #469

### CODE REVIEW

File: python/src/aiconfig/model_parser.py [hunk 1]
```diff
@@ -78,6 +79,40 @@ async def run(
...
+    ) -> list["AIConfigRuntime"]:
+        """
+        Run inference on multiple sets of parameters, one set at a time concurrently.
+        Default implementation for the run_batch method. Model Parsers may choose to override this method if they need to implement custom batch execution logic.
+        For each dictionary of parameters in `params_list``, the `run` method is invoked. All iterations are separate as we use a deep copy of `aiconfig` in each iteration.
+
+        Args:
+            prompt (Prompt): The prompt for running the inference
+            aiconfig (AIConfigRuntime): The AIConfig object containing all necessary configurations (prompts and parameters) for running the inference.
+            params_list (list[dict[str, Any]]): A List of dictionaries, where each dictionary is a set of parameters that directly influence the behaviour of inference.
```
Comment (saqadri) [inline python/src/aiconfig/model_parser.py:98]: parameters_list

File: python/src/aiconfig/model_parser.py [hunk 2]
```diff
@@ -78,6 +79,40 @@ async def run(
...
+        """
+        # Intiailize an empty list to hold the results
+        inference_results = []
+
+        # Create a deep copy of the aiconfig object to prevent mutations that could affect other iterations
+        aiconfig_deep_copy = copy.deepcopy(aiconfig)
+
+        for params in parameters_list:
+            await self.run(prompt, aiconfig_deep_copy, options, params, **kwargs)
+            inference_results.append(aiconfig_deep_copy)
```
Comment (saqadri) [inline python/src/aiconfig/model_parser.py:113]: I think we need to persist the parameters that the prompt was run with into the aiconfig, right?

File: python/src/aiconfig/model_parser.py [hunk 3]
```diff
@@ -78,6 +79,40 @@ async def run(
...
+            list[AIConfigRuntime]: A list of AIConfigRuntime objects. Each object contains the state of the AIConfigRuntime after each run using the corresponding parameter set from params_list.
+        """
+        # Intiailize an empty list to hold the results
+        inference_results = []
+
+        # Create a deep copy of the aiconfig object to prevent mutations that could affect other iterations
+        aiconfig_deep_copy = copy.deepcopy(aiconfig)
+
+        for params in parameters_list:
+            await self.run(prompt, aiconfig_deep_copy, options, params, **kwargs)
```
Comment (saqadri) [inline python/src/aiconfig/model_parser.py:112]: Why is this done serially? Shouldn't we run this in parallel across all of them?

File: python/src/aiconfig/model_parser.py [hunk 4]
```diff
@@ -78,6 +79,40 @@ async def run(
...
+        Returns:
+            list[AIConfigRuntime]: A list of AIConfigRuntime objects. Each object contains the state of the AIConfigRuntime after each run using the corresponding parameter set from params_list.
+        """
+        # Intiailize an empty list to hold the results
+        inference_results = []
+
+        # Create a deep copy of the aiconfig object to prevent mutations that could affect other iterations
+        aiconfig_deep_copy = copy.deepcopy(aiconfig)
+
+        for params in parameters_list:
```
Comment (saqadri) [inline python/src/aiconfig/model_parser.py:111]: We also need exponential back off... can come in a diff above but it's critical if run() fails.

File: python/src/aiconfig/Config.py [hunk 5]
```diff
@@ -274,6 +274,74 @@ async def run(
...
+            self,
+            parameters_list,
+            options,
+            callback_manager=self.callback_manager,
+            **kwargs,
+        )
+
+        # Extract the prompts from the result
+        prompts = []
+        for aiconfig in batch_results:
```
Comment (saqadri) [inline python/src/aiconfig/Config.py:337]: Same problem as other function that we have no idea which parameters were applied to get this particular aiconfig.

Alternatively you could do prompt.resolve with the parameters  and return the fully resolved prompt back.

**Review by saqadri — CHANGES_REQUESTED** [review]
Some comments

**Comment by Ankush-lastmile** [comment]
Added async execution. Although the current model parsers all use synchronous api calls. Something to keep in mind and update forward

**Comment by Ankush-lastmile** [comment]
>we have no idea which parameters were applied to get this particular aiconfig.

Updated output of `batch_execute()` to be a list of Execution Results, along with the corresponding completion params.

File: python/src/aiconfig/Config.py [hunk 6]
```diff
@@ -274,6 +265,77 @@ async def run(
         await self.callback_manager.run_callbacks(event)
         return response
 
+    async def run_batch(
+        self,
+        prompt_name: str,
+        parameters_list: list[dict[str, Any]],
+        options: Optional[InferenceOptions] = None,
+        **kwargs,
+    ) -> list[Tuple["ExecuteResult", JSONObject | Any]]:
```
Comment (saqadri) [inline python/src/aiconfig/Config.py:274]: Just return the corresponding params for that run in the Tuple as well.
So Tuple["ExecuteResult", JSONObject | Any, Dict]

**Review by saqadri — APPROVED** [review]
Looks good, just one comment, and then ship it

**Comment by Ankush-lastmile** [comment]
> So Tuple["ExecuteResult", JSONObject | Any, Dict]

Updated to Tuple["ExecuteResult", JSONObject | Any, Dict]

looks like some formatting changes made it into the commit .

**Review by saqadri — APPROVED** [review]
Ship it!

**Review by saqadri — APPROVED** [review]
Ship it!

**Review by saqadri — APPROVED** [review]
Ship it!

**Comment by saqadri** [comment]
lol no idea why it pasted my comment 3 times

**Pull request merged**